File: Face Invaders/src/com/cs281/face/invaders/AlienSprite.java
package com.cs281.face.invaders;

import android.graphics.*;
import java.util.Random;


public class AlienSprite extends Sprite {
	
	// Constructor
	public AlienSprite(Bitmap bmp, Rect rcBounds, BOUNDSACTION baBoundsAction)
	{
		super(bmp, rcBounds, baBoundsAction);
	}
	
	
	// General Methods
	public short Update()
	{
		// Call the base sprite Update() method
		short saSpriteAction;
		saSpriteAction = super.Update();
		
		// See if the alien should fire a missile
		Random rand = new Random();
		if (rand.nextInt(MainActivity.gDifficulty / 2) == 0)
		{
			saSpriteAction |= SA_ADDSPRITE;
		}
		
		return saSpriteAction;
	}
	
	public Sprite AddSprite()
	{	
		// Create a new missile sprite
		Rect rcBounds = new Rect(0, 0, 640, 410);
		Rect rcPos = GetPosition();
		
		Sprite sprite = null;
		if (GetBitmap() == MainActivity.gBlobboBitmap)
		{
			// Blobbo missile
			sprite = new Sprite(MainActivity.gBMissileBitmap, rcBounds, 
								BOUNDSACTION.BA_DIE);
			sprite.SetVelocity(0, 7);
		}
		else if (GetBitmap() == MainActivity.gJellyBitmap)
		{
			// Jelly missile
			sprite = new Sprite(MainActivity.gJMissileBitmap, rcBounds, 
								BOUNDSACTION.BA_DIE);
			sprite.SetVelocity(0, 5);
		}
		else
		{
			// Timmy missile
			sprite = new Sprite(MainActivity.gTMissileBitmap, rcBounds, 
					            BOUNDSACTION.BA_DIE);
			sprite.SetVelocity(0, 3);
		}
		
		// Set the missile sprite's position and return it
		sprite.SetPosition(rcPos.left + (GetWidth() / 2), rcPos.bottom);
		
		return sprite;
	}
}

File: SpaceOut 4/AlienSprite.cpp
//-----------------------------------------------------------------
// Alien Sprite Object
// C++ Source - AlienSprite.cpp
//-----------------------------------------------------------------

//-----------------------------------------------------------------
// Include Files
//-----------------------------------------------------------------
#include "AlienSprite.h"

//-----------------------------------------------------------------
// External Global Variables
//-----------------------------------------------------------------
extern Bitmap* _pBlobboBitmap;
extern Bitmap* _pBlobboBitmapf;
extern Bitmap* _pBMissileBitmap;
extern Bitmap* _pJellyBitmap;
extern Bitmap* _pJellyBitmapf;
extern Bitmap* _pJMissileBitmap;
extern Bitmap* _pTimmyBitmap;
extern Bitmap* _pTMissileBitmap;
extern int     _iDifficulty;

//-----------------------------------------------------------------
// AlienSprite Constructor(s)/Destructor
//-----------------------------------------------------------------
AlienSprite::AlienSprite(Bitmap* pBitmap, RECT& rcBounds,
  BOUNDSACTION baBoundsAction, Bitmap* pBitmapf) : Sprite(pBitmap, rcBounds,
  baBoundsAction)
{
	m_pBitmapf =pBitmapf;
}

AlienSprite::~AlienSprite()
{
}

//-----------------------------------------------------------------
// AlienSprite General Methods
//-----------------------------------------------------------------
SPRITEACTION AlienSprite::Update()
{
  // Call the base sprite Update() method
  SPRITEACTION saSpriteAction;
  saSpriteAction = Sprite::Update();

  // See if the alien should fire a missile
  if ((rand() % (_iDifficulty / 2)) == 0)
    saSpriteAction |= SA_ADDSPRITE;

  return saSpriteAction;
}

Sprite* AlienSprite::AddSprite()
{
  // Create a new missile sprite
  RECT    rcBounds = { 0, 0, 640, 410 };
  RECT    rcPos = GetPosition();
  Sprite* pSprite = NULL;
  if (GetBitmap() == _pBlobboBitmap)
  {
    // Blobbo missile
    pSprite = new Sprite(_pBMissileBitmap, rcBounds, BA_DIE);
    pSprite->SetVelocity(0, 7);
  }
  else if (GetBitmap() == _pJellyBitmap)
  {
    // Jelly missile
    pSprite = new Sprite(_pJMissileBitmap, rcBounds, BA_DIE);
    pSprite->SetVelocity(0, 5);
  }
  else
  {
    // Timmy missile
    pSprite = new Sprite(_pTMissileBitmap, rcBounds, BA_DIE);
    pSprite->SetVelocity(0, 3);
  }

  // Set the missile sprite's position and return it
  pSprite->SetPosition(rcPos.left + (GetWidth() / 2), rcPos.bottom);
  return pSprite;
}

void AlienSprite::Flip(){
	  if (GetBitmap() == _pBlobboBitmap || GetBitmap() == _pBlobboBitmapf)
	  {
			Bitmap* tmp = m_pBitmapf;
			m_pBitmapf = m_pBitmap;
			m_pBitmap = tmp;
	  }
	  else if (GetBitmap() == _pJellyBitmap || GetBitmap() == _pJellyBitmapf )
	  {
			Bitmap* tmp = m_pBitmapf;
			m_pBitmapf = m_pBitmap;
			m_pBitmap = tmp;
		  //jelly needs to be flipped
	  }
	  else{
		//timmy won't need to be flipped
	  }
}  
